// SPDX-License-Identifier: Apache-2.0
syntax = "proto3";
package loom;
option optimize_for = LITE_RUNTIME;

// Enum member name-to-value mapping (for symbolic display)
message EnumMember {
  string name = 1;     // e.g. "StIdle"
  uint64 value = 2;    // e.g. 0
}

// One per FF/register — already grouped at compile time
message ScanVariable {
  string name = 1;     // original HDL path from hdlname attr, dot-separated
  uint32 width = 2;    // total bit width of the variable
  uint32 offset = 3;   // starting chain position (bits [offset..offset+width-1])
  repeated EnumMember enum_members = 4;  // empty if not an enum type
  bytes reset_value = 5;  // LE-packed reset value (empty = no reset / default 0)
}

// Reset DPI mapping: func_id → scan chain position
// Used to inject DPI return values into the scan image before scan-in
message ResetDpiMapping {
  uint32 func_id = 1;       // dispatch table index
  uint32 scan_offset = 2;   // bit position in scan chain
  uint32 scan_width = 3;    // bit width in scan chain
}

message ScanMap {
  uint32 chain_length = 1;
  repeated ScanVariable variables = 2;
  bytes initial_scan_image = 3;  // pre-built full chain blob for scan-based init
  repeated ResetDpiMapping reset_dpi_mappings = 4;  // reset DPI → scan position mappings
}

// One per shadow-ported memory
message MemoryEntry {
  string name = 1;           // HDL path (dot-separated)
  uint32 depth = 2;          // number of entries
  uint32 width = 3;          // bits per entry
  uint32 addr_bits = 4;      // ceil_log2(depth)
  uint32 base_addr = 5;      // byte offset in unified address space
  uint32 end_addr = 6;       // exclusive byte end
  bytes  initial_content = 7; // LE-packed inline init (from initial assignments)
  string init_file = 8;      // path to $readmemh/$readmemb file (resolved at runtime)
  bool   init_file_hex = 9;  // true = $readmemh, false = $readmemb
}

message MemMap {
  uint32 total_bytes = 1;    // address space size
  uint32 addr_bits = 2;      // global address width
  uint32 data_bits = 3;      // max data width
  uint32 num_memories = 4;
  repeated MemoryEntry memories = 5;
}

// Single state snapshot (dump file)
message Snapshot {
  uint64 cycle_count = 1;
  uint64 dut_time = 2;
  uint32 design_id = 3;
  bytes  raw_scan_data = 4;        // raw chain words packed LE
  ScanMap scan_map = 5;            // embedded — makes file self-contained
  MemMap mem_map = 6;              // memory map metadata
  bytes raw_mem_data = 7;          // concatenated memory dump
}
